azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -59,95 +62,158 @@ def __init__(self, wire_client, full_goal_state=False, base_incarnation=None):
         """
         uri = GOAL_STATE_URI.format(wire_client.get_endpoint())
         self.xml_text = wire_client.fetch_config(uri, wire_client.get_header())
-        xml_doc = parse_doc(self.xml_text)
-
-        self.incarnation = findtext(xml_doc, "Incarnation")
-        self.expected_state = findtext(xml_doc, "ExpectedState")
-        role_instance = find(xml_doc, "RoleInstance")
+        self._xml_doc = parse_doc(self.xml_text)
+
+        self._ext_config_retriever = ext_config_retriever
+        self._wire_client = wire_client
+
+        # Cached properties
+        # Implementation note: cached_property was not used because it considers a None value to mean the property
+        # wasn't retrieved
+        self._ext_conf = None
+        self._ext_conf_retrieved = False
```

Comment:
the use of the "_retrieved" variables threw me off; seems like a check for _ext_conf == None would be the same

Reply:
We need the "_retrieved" values for the case when the object we're retrieving is None. For example, if the ExtensionsConfig element isn't in the GoalState xml, then we'll set it to None. The _retrieved property prevents us from doing this check over and over. It also applies in the case where we can't retrieve the document for any reason.